### ISSUE TITLE
email texts: line-breaks substitutions inconsistent

### ISSUE THREAD

**Issue by torvista** [zencart/zencart#2742]
Is this really the dog's breakfast it appears to be?
https://www.zen-cart.com/showthread.php?225812-email-content-line-breaks-for-layout-formatting

**Comment by wiztechinc** [comment]
yup, that's what I was complaining about the editing of emails. Messy and even bizarre

**Comment by torvista** [comment]
ok. So what's the plan?
I'm working on a preview mod that I really do not want to work round this mess as it is.

Why not continue to use \n or multiples of, line breaks in the constants, 
use them as is in the text emails, 
load them as is into the html blocks array, then nl2br the blocks in the array just prior to sending?
I don't see why a single \n should be removed...it should be changed to br, no?

**Comment by wiztechinc** [comment]
br is html - the mixture of the two makes no sense to me and may be a problem. I do know that the plain text emails need to be very very simple. Email programs are the end result - no matter what you send, the email program displays its own way. That said, I'm wondering if this is something that needs to be fully researched today. Unless there is someone who is presently working with this type of thing regularly beyond Zen Cart who is watching this project and wants to chime in. And yes, I volunteer to research if need be.

**Comment by torvista** [comment]
So, just looking at create_account, there are three variants of line-break substitution, followed by more in function zen_email. 
What is the end-plan for this...so someone can think about a solution?

**Comment by torvista** [comment]
This requires a comment to guide a PR, or not.

**Comment by drbyte** [comment]
The end-result is simple:
- plain-text emails can't handle a `<br>`, because plain-text email clients will just display it as text, not as a line-break. So, ALL line-breaks should be `"\n"`
- rendering of the content for HTML purposes (all $html_msg array entries) needs to `nl2br()` things in order to convert those `\n` to `<br>` tags.
- the double-line-breaks denote a paragraph change, because as you know a single line-break at the end of a line does not insert a "blank line", so will not insert enough "space" to denote a change in thought like a paragraph would, particularly when converted to HTML; hence the use of double-line-breaks.
- several of the single/multiple `\n` in existing definitions are a result of extensive testing across a dozen clients and all browsers available when the feature was initially implemented. Of course that cannot account for all the changes a merchant might make.


NOTE: IT'S IMPORTANT TO NOTE THAT ALL EMAIL CLIENTS STILL ADHERE BEST TO XHTML4 SPEC AS COMMON DENOMINATOR. YOU SHOULD ALWAYS EXPECT THAT YOUR END-USERS CAN ONLY READ VERY OLD HTML FORMAT, AND NOT DEPEND ON ANY SUPPORT FOR MODERN HTML5 SYNTAX. Even most recent versions of Outlook can't consistently display all HTML5 markup. It's sad that still in 2020 the vast majority of email clients *need to use table-based layouts, as they are incapable of properly rendering modern approaches to HTML layout in a way that looks correct to the end-user.

**Comment by torvista** [comment]
Ok, so this replacement is wrong as I assumed in my forum post:
>For the html-version of the email, the double line-breaks "\n\n" are replaced by an xhtml line break < br />, and the remaining single line-breaks \n are removed completely.

**Comment by drbyte** [comment]
> Ok, so this replacement is wrong as I assumed in my forum post:

In some cases, maybe, yes.
